src/java.base/share/classes/jdk/internal/foreign/SegmentBulkOperations.java

```diff
@@ -57,46 +57,57 @@ private SegmentBulkOperations() {}
+     * Said tail scheme is simpler and faster than trying to apply a combination of
+     * operations with different units of memory to clean up the tail.
+     * <p>
+     * This method is using Java for segments that are smaller than the configurable
+     * {@code NATIVE_THRESHOLD_FILL} meaning it does not have to pay the price for
+     * transitioning to native code.
+     *
+     * @param dst   segment to fill
+     * @param value to fill the segment with
+     * @return the provided {@code dst} segment
+     */
     @ForceInline
     public static MemorySegment fill(AbstractMemorySegmentImpl dst, byte value) {
         dst.checkReadOnly(false);
-        if (dst.length == 0) {
-            // Implicit state check
-            dst.sessionImpl().checkValidState();
-        } else if (dst.length < NATIVE_THRESHOLD_FILL) {
-            // 0 <= length < FILL_NATIVE_LIMIT : 0...0X...XXXX
-
-            // Handle smaller segments directly without transitioning to native code
-            final long u = Byte.toUnsignedLong(value);
-            final long longValue = u << 56 | u << 48 | u << 40 | u << 32 | u << 24 | u << 16 | u << 8 | u;
+        final long len = dst.length;
+        /* The multiplication below is equivalent to:
+             long u = Byte.toUnsignedLong(value);
+             long longValue = u << 56 | u << 48 | u << 40 | u << 32 | u << 24 | u << 16 | u << 8 | u; */
+        final long longValue = 0x01010101_01010101L * Byte.toUnsignedLong(value);
 
-            int offset = 0;
-            // 0...0X...X000
-            final int limit = (int) (dst.length & (NATIVE_THRESHOLD_FILL - 8));
-            for (; offset < limit; offset += Long.BYTES) {
-                SCOPED_MEMORY_ACCESS.putLongUnaligned(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + offset, longValue, !Architecture.isLittleEndian());
-            }
-            int remaining = (int) dst.length - limit;
-            // 0...0X00
-            if (remaining >= Integer.BYTES) {
-                SCOPED_MEMORY_ACCESS.putIntUnaligned(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + offset, (int) longValue, !Architecture.isLittleEndian());
-                offset += Integer.BYTES;
-                remaining -= Integer.BYTES;
+        // Switch on log2(len) = 64 - Long.numberOfLeadingZeros(len)
+        switch (64 - Long.numberOfLeadingZeros(len)) {
+            case 0 -> dst.sessionImpl().checkValidState(); // Implicit state check
+            case 1 -> SCOPED_MEMORY_ACCESS.putByte(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset(), value);
+            case 2 -> {
+                SCOPED_MEMORY_ACCESS.putShortUnaligned(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset(), (short) longValue, !Architecture.isLittleEndian());
+                SCOPED_MEMORY_ACCESS.putShortUnaligned(dst.sessionImpl(), dst.unsafeGetBase(), dst.unsafeGetOffset() + len - Short.BYTES, (short) longValue, !Architecture.isLittleEndian());
```

Comment:
So that I understand, since this configuration might pick up both length = 2 and length = 3, in 50% of cases there's a redundant store? This is also true for length = 4, 5, 6, 7 -- but in this case the chance of redundant store is reduced to 25%.

Reply:
That is correct. But it seems the hardware can handle redundant stores better than handling more branching. One idea is to add in "halfbits" to the tableswitch which would reduce redundant stores. I think this would have little to no performance effect while complicating the code, but it might be worth trying anyhow.